pkg/controller/user-cluster-controller-manager/constraint-template-controller/controller.go

```diff
@@ -0,0 +1,152 @@
+
+	"github.com/open-policy-agent/frameworks/constraint/pkg/apis/templates/v1beta1"
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	v1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// This controller syncs the kubermatic constraint templates to gatekeeper constraint templates on the user cluster.
+	controllerName = "constraint_template_controller"
+)
+
+type reconciler struct {
+	ctx        context.Context
+	log        *zap.SugaredLogger
+	userClient ctrlruntimeclient.Client
+	seedClient ctrlruntimeclient.Client
+	recorder   record.EventRecorder
+}
+
+func Add(ctx context.Context, log *zap.SugaredLogger, userMgr, seedMgr manager.Manager) error {
+	log = log.Named(controllerName)
+
+	r := &reconciler{
+		ctx:        ctx,
+		log:        log,
+		userClient: userMgr.GetClient(),
+		seedClient: seedMgr.GetClient(),
+		recorder:   userMgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.New(controllerName, userMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	ctSource := &source.Kind{Type: &v1.ConstraintTemplate{}}
+	if err := ctSource.InjectCache(seedMgr.GetCache()); err != nil {
+		return fmt.Errorf("failed to inject seed cache into watch: %v", err)
+	}
+
+	// Watch for changes to ConstraintTemplates
+	if err = c.Watch(ctSource, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to establish watch for the ConstraintTemplates %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("ConstraintTemplate", request.Name)
+	log.Debug("Reconciling")
+
+	constraintTemplate := &v1.ConstraintTemplate{}
+	if err := r.seedClient.Get(r.ctx, request.NamespacedName, constraintTemplate); err != nil {
+		if kerrors.IsNotFound(err) {
+			log.Debug("constraint template not found, returning")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get constraint template: %v", err)
+	}
+
+	if constraintTemplate.DeletionTimestamp != nil {
+		if !kuberneteshelper.HasFinalizer(constraintTemplate, kubermaticapiv1.ConstraintTemplateCleanupFinalizer) {
+			return reconcile.Result{}, nil
+		}
+		if err := r.userClient.Delete(r.ctx, &v1beta1.ConstraintTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: constraintTemplate.Name,
```

Comment:
Please add the namespace of the yet to be deleted resource.

Reply:
its cluster scoped so no need here